[lldb] Mark scripted frames as synthetic instead of artificial

medismailben opened the pull request:
This patch changes the way frames created from scripted affordances like Scripted Threads are displayed. Currently, they're marked artificial which is used usually for compiler generated frames.

This patch changes that behaviour by introducing a new synthetic StackFrame kind and moves 'artificial' to be a distinct StackFrame attribut.

On top of making these frames less confusing, this allows us to know when a frame was created from a scripted affordance.

rdar://155949703

jimingham commented:
Right.  Seems to me there are two orthogonal notions here: the provider of the frame info: Historical, Unwind Based, Synthetic (we should also include "Inferred" for instance the insertion of tail call frames is inferred); and the language based characterization of the frames (Inline, Artificial).

jimingham commented on lldb/include/lldb/Target/StackFrame.h:
```diff
@@ -567,6 +571,10 @@ class StackFrame : public ExecutionContextScope,
+  /// Is this an artificial stack frame (e.g. a synthesized result of inferring
+  /// missing tail call frames from a backtrace) with limited support for
+  /// local variables.
+  bool m_artificial;
```
Note, we also use "artificial" to mean the frame is a language-generated frame that most users probably don't want to see.  There's been a suggestion that lldb should auto-hide and auto-step through this sort of "artificial" frame.  But this is very much NOT how we want to treat the "missing tail call frames".  So they need a different flag to indicate what they are.
They really should be "historical" frames - like other historical frames they are just a PC & CFA, but their stack doesn't currently exist - it's been overwritten by the function it tail-called to.
Anyway, you should change the comment here to refer to the more common usage...

JDevlieghere reviewed (commented):
Reading the discussion, my main feedback is:
 
- Add a comment to `StackFrame::Kind` to explain that it's about where a stack frame comes from.
- Add a comment to `m_artificial` to explain that it's orthogonal to the kind. Also right now it has the same comment as `Kind::Synthetic` which can't be right. 
- Rather than a format entity for whether a frame is synthetic, it seems that we should have a format entity that prints the kind (which could be empty for a regular frame and print `[historical]` and `[synthetic]` respectively

medismailben committed:
[lldb] Mark scripted frames as synthetic instead of artificial

This patch changes the way frames created from scripted affordances like
Scripted Threads are displayed. Currently, they're marked artificial
which is used usually for compiler generated frames.

This patch changes that behaviour by introducing a new synthetic
StackFrame kind and moves 'artificial' to be a distinct StackFrame attribut.

On top of making these frames less confusing, this allows us to know
when a frame was created from a scripted affordance.

rdar://155949703

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben commented:
@JDevlieghere @jimingham I think I've address all the comments here, do you mind taking another look ?

JDevlieghere reviewed (approved):
LGTM

jimingham commented:
LGTM2
